app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,489 @@
+    return fileUriArrayList;
+  }
+
+  public int getTotalFilesForTransfer() {
+    return totalFilesForTransfer;
+  }
+
+  public void setTotalFilesForTransfer(int totalFilesForTransfer) {
+    this.totalFilesForTransfer = totalFilesForTransfer;
+  }
+
+  public @NonNull ArrayList<FileItem> getFilesForTransfer() {
+    return filesForTransfer;
+  }
+
+  public void setFilesForTransfer(@NonNull ArrayList<FileItem> fileItems) {
+    this.filesForTransfer = fileItems;
+  }
+
+  public void incrementTotalFilesSent() {
+    this.filesSent++;
+  }
+
+  public boolean allFilesSent() {
+    return (filesSent == totalFilesForTransfer);
+  }
+
+  public @NonNull String getZimStorageRootPath() {
+    return (sharedPreferenceUtil.getPrefStorage() + "/Kiwix/");
+  }
+
+  public @NonNull InetAddress getFileReceiverDeviceAddress() {
+    return fileReceiverDeviceAddress;
+  }
+
+  public static void copyToOutputStream(@NonNull InputStream inputStream, @NonNull
+      OutputStream outputStream)
+      throws IOException {
+    byte[] bufferForBytes = new byte[1024];
+    int bytesRead;
+
+    Log.d(TAG, "Copying to OutputStream...");
+    while ((bytesRead = inputStream.read(bufferForBytes)) != -1) {
+      outputStream.write(bufferForBytes, 0, bytesRead);
+    }
+
+    outputStream.close();
+    inputStream.close();
+    Log.d(LocalFileTransferActivity.TAG, "Both streams closed");
+  }
+
+  public void setClientAddress(@Nullable InetAddress clientAddress) {
+    if (clientAddress == null) {
+      // null is returned only in case of a failed handshake
+      displayToast(R.string.device_not_cooperating, Toast.LENGTH_LONG);
+      ((Callbacks) activity).onFileTransferComplete(false);
+      return;
+    }
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress;
+    startFileTransfer();
+  }
+
+  private void startFileTransfer() {
+    if (isGroupFormed() && !isFileSender) {
+      ((Callbacks) activity).onFilesForTransferAvailable(filesForTransfer);
+
+      receiverDeviceAsyncTask = new ReceiverDeviceAsyncTask(this);
+      receiverDeviceAsyncTask.execute();
+
+    } else if (isGroupFormed()) { // && isFileSender
+      {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer");
+
+        fileReceiverDeviceAddress =
+            (isGroupOwner()) ? selectedPeerDeviceInetAddress : getGroupOwnerAddress();
+
+        displayToast(R.string.preparing_files, Toast.LENGTH_LONG);
+        senderDeviceAsyncTaskArray = new SenderDeviceAsyncTask(this, activity);
+        senderDeviceAsyncTaskArray.execute(fileUriArrayList.toArray(new Uri[0]));
+      }
+    }
+  }
+
+  public void changeStatus(int itemIndex, @FileItem.FileStatus int status) {
+    filesForTransfer.get(itemIndex).setFileStatus(status);
+    ((Callbacks) activity).onFileStatusChanged(itemIndex);
+  }
+
+  private void cancelAsyncTasks() {
+    if (peerGroupHandshakeAsyncTask != null) {
```

Comment:
you could make a vararg method here so you don't repeat yourself.

```
cancelIfNotNull(peerGroupHandshakeAsyncTask, senderDeviceAsyncTaskArray, receiverDeviceAsyncTask)

private void cancelIfNotNull(Asynctask... tasks){
  for (aTask: tasks) {
    if (aTask != null) {
       atask.cancel()
     }
  }
}
```

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/d326deb4f301a4231bbe2de936beb7288967ca55)